aerofiles/seeyou/reader.py

```diff
@@ -193,3 +193,90 @@ def decode_description(self, description):
             return None
 
         return description
+
+    def decode_taskname(self, fields):
+        if not fields:
+            raise ParserError('Name field must not be empty')
+
+        return fields[0]
+
+    def decode_tasklist(self, fields, waypoints):
```

Comment:
> the next() method provides the logic for the next "element" in the object (in the case of a cup file with only waypoints, this method gives the next waypoint). in the case of a cup file with waypoints and a task, this handling is not yet covered.

correct

> the call to the methods decode_tasklist(), decode_taskname, decode_options, decode_taskpoint could be incorporated in some way into this next() method to ultimately provide the task back to the user.

> the __iter__() method seems a bit counter intuitive if the object contains both waypoints and a task. what are your thoughts on this?

I agree on that. In retrospect implementing this as an iterator was probably not such a great idea. Using iterators is great if you have very large files like a CSV file with millions of lines, but since CUP files tend to be rather small (< 10k lines) this is probably not worth it.

the way forward as I would imagine it:

- while introducing new API we should try to stay compatible with the current API
- end goal: `Reader().read('path/to/some/file.cup')` will return a dict with `waypoints` and `tasks`
- add a `read(path)` method roughly like this
```python
def read(self, fp):
    waypoints = filter(map(self.decode_waypoint, csv.reader(fp)))
    return dict(waypoints=waypoints)
```
- use `waypoints = self.read(self.fp)['waypoints']` in `next()` and just iterate/yield those waypoints
- refactor `read()` to also read tasks